File: android/src/main/java/com/lufinkey/react/eventemitter/RNEventEmitter.java

package com.lufinkey.react.eventemitter;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.Callback;
import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.WritableArray;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.module.annotations.ReactModule;
import com.facebook.react.modules.core.DeviceEventManagerModule;

import java.util.HashMap;

@ReactModule(name = "RNEventEmitter")
public class RNEventEmitter extends ReactContextBaseJavaModule {
	private final ReactApplicationContext reactContext;

	public static final String EVENT_NAME = "ayylmao_dicksnshit_nobodyUsethisevent PLS OK THANKS";

	private final HashMap<Integer, RNEventConformer> registeredModules = new HashMap<>();
	private final HashMap<Integer, RNModuleEvents> modules = new HashMap<>();

	public RNEventEmitter(ReactApplicationContext reactContext) {
		super(reactContext);
		this.reactContext = reactContext;
	}

	@Override
	public String getName() {
		return "RNEventEmitter";
	}



	public static RNEventEmitter getMainEventEmitter(ReactApplicationContext context) {
		return context.getNativeModule(RNEventEmitter.class);
	}



	public static void registerEventEmitterModule(ReactApplicationContext context, int moduleId, RNEventConformer module) {
		RNEventEmitter eventEmitter = getMainEventEmitter(context);
		if(eventEmitter == null) {
			System.out.println("Error: No RNEventEmitter is available to register to");
			return;
		}
		eventEmitter.registerModule(moduleId, module);
	}

	private void registerModule(int moduleId, RNEventConformer module) {
		synchronized (registeredModules) {
			if(registeredModules.containsValue(module)) {
				throw new IllegalArgumentException("Module "+module+" has already been registered");
			}
			if(registeredModules.containsKey(moduleId)) {
				throw new IllegalArgumentException("moduleId "+moduleId+" has already been registered to a module");
			}
			registeredModules.put(moduleId, module);
		}
	}



	private RNModuleEvents getModuleEvents(int moduleId) {
		synchronized (modules) {
			RNModuleEvents moduleEvents = modules.get(moduleId);
			if (moduleEvents == null) {
				moduleEvents = new RNModuleEvents();
				modules.put(moduleId, moduleEvents);
			}
			return moduleEvents;
		}
	}

	private RNEventConformer getRegisteredModule(int moduleId) {
		synchronized (registeredModules) {
			return registeredModules.get(moduleId);
		}
	}

	private Integer getRegisteredModuleID(RNEventConformer module) {
		synchronized (registeredModules) {
			for (HashMap.Entry<Integer, RNEventConformer> entry : registeredModules.entrySet()) {
				if(entry.getValue() == module) {
					return entry.getKey();
				}
			}
			return null;
		}
	}

	private WritableArray fromObjectArray(Object[] args) {
		WritableArray array = Arguments.createArray();
		for(Object arg : args) {
			if(arg == null) {
				array.pushNull();
			}
			else if(arg instanceof Boolean) {
				array.pushBoolean((Boolean)arg);
			}
			else if(arg instanceof Integer) {
				array.pushInt((Integer)arg);
			}
			else if(arg instanceof Double) {
				array.pushDouble((Double)arg);
			}
			else if(arg instanceof Float) {
				array.pushDouble((double)((Float)arg));
			}
			else if(arg instanceof String) {
				array.pushString((String)arg);
			}
			else if(arg instanceof WritableArray) {
				array.pushArray((WritableArray)arg);
			}
			else if(arg instanceof WritableMap) {
				array.pushMap((WritableMap)arg);
			}
			else {
				throw new IllegalArgumentException("Illegal object type");
			}
		}
		return array;
	}



	public static void emitEvent(ReactApplicationContext context, RNEventConformer module, String eventName, Object... args) {
		RNEventEmitter eventEmitter = getMainEventEmitter(context);
		if(eventEmitter == null) {
			System.out.println("Error: No RNEventEmitter is available to emit "+eventName+" event");
			return;
		}
		eventEmitter.emit(module, eventName, args);
	}

	public void emit(RNEventConformer module, String eventName, Object... args) {
		Integer moduleId = getRegisteredModuleID(module);
		if(moduleId == null) {
			System.out.println("Error: Cannot emit "+eventName+" event before "+module+" module has been registered");
			return;
		}
		getModuleEvents(moduleId).emit(eventName, args);

		WritableMap jsEvent = Arguments.createMap();
		jsEvent.putInt("moduleId", moduleId);
		jsEvent.putString("eventName", eventName);
		jsEvent.putArray("args", fromObjectArray(args));

		reactContext.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class).emit(EVENT_NAME, jsEvent);

		module.onEvent(eventName, args);
		module.onNativeEvent(eventName, args);
	}

	@ReactMethod
	private void emit(int moduleId, String eventName, ReadableArray args) {
		RNEventConformer module = getRegisteredModule(moduleId);
		if(module == null) {
			throw new IllegalArgumentException("No module registered with ID "+moduleId);
		}

		Object[] argsArray = args.toArrayList().toArray();
		getModuleEvents(moduleId).emit(eventName, argsArray);

		module.onEvent(eventName, argsArray);
		module.onJSEvent(eventName, argsArray);
	}



	public void addListener(RNEventConformer module, String eventName, Callback callback, boolean once) {
		Integer moduleId = getRegisteredModuleID(module);
		if(moduleId == null) {
			throw new IllegalArgumentException("Module "+module+" has not been registered");
		}
		addListener(moduleId, eventName, callback, once);
	}

	public void addListener(RNEventConformer module, String eventName, Callback callback) {
		addListener(module, eventName, callback, false);
	}

	public void addListener(int moduleId, String eventName, Callback callback) {
		addListener(moduleId, eventName, callback, false);
	}

	public void addListener(int moduleId, String eventName, Callback callback, boolean once) {
		getModuleEvents(moduleId).addListener(eventName, callback, once);
	}



	public void prependListener(RNEventConformer module, String eventName, Callback callback, boolean once) {
		Integer moduleId = getRegisteredModuleID(module);
		if(moduleId == null) {
			throw new IllegalArgumentException("Module "+module+" has not been registered");
		}
		prependListener(moduleId, eventName, callback, once);
	}

	public void prependListener(RNEventConformer module, String eventName, Callback callback) {
		prependListener(module, eventName, callback, false);
	}

	public void prependListener(int moduleId, String eventName, Callback callback) {
		prependListener(moduleId, eventName, callback, false);
	}

	public void prependListener(int moduleId, String eventName, Callback callback, boolean once) {
		getModuleEvents(moduleId).prependListener(eventName, callback, once);
	}



	public void removeListener(RNEventConformer module, String eventName, Callback callback) {
		Integer moduleId = getRegisteredModuleID(module);
		if(moduleId == null) {
			throw new IllegalArgumentException("Module "+module+" has not been registered");
		}
		removeListener(moduleId, eventName, callback);
	}

	public void removeListener(int moduleId, String eventName, Callback callback) {
		getModuleEvents(moduleId).removeListener(eventName, callback);
	}



	public void removeAllListeners(RNEventConformer module, String eventName) {
		Integer moduleId = getRegisteredModuleID(module);
		if(moduleId == null) {
			throw new IllegalArgumentException("Module "+module+" has not been registered");
		}
		removeAllListeners(moduleId, eventName);
	}

	public void removeAllListeners(int moduleId, String eventName) {
		getModuleEvents(moduleId).removeAllListeners(eventName);
	}
}

File: ios/RNEventEmitter.m

#import "RNEventEmitter.h"
#import "RNEventConformer.h"
#import "RNModuleEvents.h"
#import "RNEventBridge.h"

@interface RNEventEmitter() {
	NSMutableDictionary<NSNumber*, id<RNEventConformer>>* _registeredModules;
	NSMutableDictionary<NSNumber*, RNModuleEvents*>* _modules;
}
-(RNModuleEvents*)moduleEventsForID:(int)moduleId;
-(NSNumber*)registeredModuleIDForModule:(id<RNEventConformer>)module;
@end


@implementation RNEventEmitter

@synthesize bridge = _bridge;

static NSString* EVENT_NAME = @"ayylmao_dicksnshit_nobodyUsethisevent PLS OK THANKS";

RCT_EXPORT_MODULE()

-(id)init {
	if(self = [super init]) {
		_registeredModules = [NSMutableDictionary dictionary];
		_modules = [NSMutableDictionary dictionary];
	}
	return self;
}

+(BOOL)requiresMainQueueSetup {
	return NO;
}

+(RNEventEmitter*)eventEmitterForBridge:(RCTBridge*)bridge {
	return [RNEventBridge moduleForClass:[RNEventEmitter class] bridge:bridge];
}



+(void)registerEventEmitterModule:(id<RNEventConformer>)module withID:(int)moduleId bridge:(RCTBridge*)bridge {
	RNEventEmitter* eventEmitter = [self eventEmitterForBridge:bridge];
	if(eventEmitter == nil) {
		NSLog(@"Error: No RNEventEmitter is available to register to");
		return;
	}
	[eventEmitter registerModule:module withID:moduleId];
}

-(void)registerModule:(id<RNEventConformer>)module withID:(int)moduleId {
	@synchronized (_registeredModules) {
		if([[_registeredModules allValues] indexOfObjectIdenticalTo:module] != NSNotFound) {
			@throw [NSException exceptionWithName:NSInvalidArgumentException reason:@"Module has already been registered" userInfo:nil];
		}
		if(_registeredModules[@(moduleId)] != nil) {
			@throw [NSException exceptionWithName:NSInvalidArgumentException reason:@"moduleId has already been registered to a module" userInfo:nil];
		}
		_registeredModules[@(moduleId)] = module;
	}
}

-(RNModuleEvents*)moduleEventsForID:(int)moduleId {
	@synchronized (_modules) {
		RNModuleEvents* moduleEvents = _modules[@(moduleId)];
		if (moduleEvents == nil) {
			moduleEvents = [[RNModuleEvents alloc] init];
			_modules[@(moduleId)] = moduleEvents;
		}
		return moduleEvents;
	}
}

-(id<RNEventConformer>)registeredModuleForID:(int)moduleId {
	@synchronized (_registeredModules) {
		return _registeredModules[@(moduleId)];
	}
}

-(NSNumber*)registeredModuleIDForModule:(id<RNEventConformer>)module {
	@synchronized (_registeredModules) {
		for (NSNumber* moduleId in [_registeredModules allKeys]) {
			id<RNEventConformer> cmpModule = _registeredModules[moduleId];
			if(cmpModule == module) {
				return moduleId;
			}
		}
		return nil;
	}
}



+(void)emitEvent:(NSString*)event withParams:(NSArray*)params module:(id<RNEventConformer>)module bridge:(RCTBridge*)bridge {
	RNEventEmitter* eventEmitter = [self eventEmitterForBridge:bridge];
	if(eventEmitter == nil) {
		NSLog(@"Error: No RNEventEmitterModule is available to emit %@ event", event);
		return;
	}
	[eventEmitter emitEvent:event withParams:params forModule:module];
}

-(void)emitEvent:(NSString*)event withParams:(NSArray*)params forModule:(id<RNEventConformer>)module {
	NSNumber* moduleId = [self registeredModuleIDForModule:module];
	if(moduleId == nil) {
		NSLog(@"Error: Cannot emit %@ event before %@ module has been registered", event, module);
		return;
	}
	[[self moduleEventsForID:moduleId.intValue] emitEvent:event withParams:params];
	
	NSMutableDictionary* jsEvent = [NSMutableDictionary dictionary];
	jsEvent[@"moduleId"] = moduleId;
	jsEvent[@"eventName"] = event;
	jsEvent[@"args"] = params;
	
	[RNEventBridge sendEvent:EVENT_NAME body:jsEvent bridge:_bridge];
	
	if([module respondsToSelector:@selector(onEvent:params:)]) {
		[module onEvent:event params:params];
	}
	if([module respondsToSelector:@selector(onNativeEvent:params:)]) {
		[module onNativeEvent:event params:params];
	}
}

RCT_EXPORT_METHOD(emit:(int)moduleId event:(NSString*)event params:(NSArray*)params) {
	id<RNEventConformer> module = [self registeredModuleForID:moduleId];
	if(module == nil) {
		@throw [NSException exceptionWithName:NSInvalidArgumentException reason:@"No module registered with that ID" userInfo:nil];
	}
	
	[[self moduleEventsForID:moduleId] emitEvent:event withParams:params];
	
	if([module respondsToSelector:@selector(onEvent:params:)]) {
		[module onEvent:event params:params];
	}
	if([module respondsToSelector:@selector(onJSEvent:params:)]) {
		[module onJSEvent:event params:params];
	}
}



-(void)addListener:(void(^)(NSArray*))listener forEvent:(NSString*)event fromModule:(id<RNEventConformer>)module runOnce:(BOOL)once {
	NSNumber* moduleId = [self registeredModuleIDForModule:module];
	if(moduleId == nil) {
		@throw [NSException exceptionWithName:NSInvalidArgumentException reason:@"Module has not been registered" userInfo:nil];
	}
	[self addListener:listener forEvent:event fromModuleWithID:moduleId.intValue runOnce:once];
}

-(void)addListener:(void(^)(NSArray*))listener forEvent:(NSString*)event fromModule:(id<RNEventConformer>)module {
	[self addListener:listener forEvent:event fromModule:module runOnce:NO];
}

-(void)addListener:(void(^)(NSArray*))listener forEvent:(NSString*)event fromModuleWithID:(int) moduleId runOnce:(BOOL)once {
	[[self moduleEventsForID:moduleId] addListener:listener forEvent:event onlyOnce:once];
}

-(void)addListener:(void(^)(NSArray*))listener forEvent:(NSString*)event fromModuleWithID:(int)moduleId {
	[self addListener:listener forEvent:event fromModuleWithID:moduleId runOnce:NO];
}



-(void)prependListener:(void(^)(NSArray*))listener forEvent:(NSString*)event fromModule:(id<RNEventConformer>)module runOnce:(BOOL)once {
	NSNumber* moduleId = [self registeredModuleIDForModule:module];
	if(moduleId == nil) {
		@throw [NSException exceptionWithName:NSInvalidArgumentException reason:@"Module has not been registered" userInfo:nil];
	}
	[self prependListener:listener forEvent:event fromModuleWithID:moduleId.intValue runOnce:once];
}

-(void)prependListener:(void(^)(NSArray*))listener forEvent:(NSString*)event fromModule:(id<RNEventConformer>)module {
	[self prependListener:listener forEvent:event fromModule:module runOnce:NO];
}

-(void)prependListener:(void(^)(NSArray*))listener forEvent:(NSString*)event fromModuleWithID:(int) moduleId runOnce:(BOOL)once {
	[[self moduleEventsForID:moduleId] addListener:listener forEvent:event onlyOnce:once];
}

-(void)prependListener:(void(^)(NSArray*))listener forEvent:(NSString*)event fromModuleWithID:(int)moduleId {
	[self addListener:listener forEvent:event fromModuleWithID:moduleId runOnce:NO];
}



-(void)removeListener:(void(^)(NSArray*))listener forEvent:(NSString*)event fromModule:(id<RNEventConformer>)module {
	NSNumber* moduleId = [self registeredModuleIDForModule:module];
	if(moduleId == nil)
	{
		@throw [NSException exceptionWithName:NSInvalidArgumentException reason:@"Module has not been registered" userInfo:nil];
	}
	[self removeListener:listener forEvent:event fromModuleWithID:moduleId.intValue];
}

-(void)removeListener:(void(^)(NSArray*))listener forEvent:(NSString*)event fromModuleWithID:(int)moduleId {
	[[self moduleEventsForID:moduleId] removeListener:listener forEvent:event];
}



-(void)removeAllListenersFromModule:(id<RNEventConformer>)module forEvent:(NSString*)event {
	NSNumber* moduleId = [self registeredModuleIDForModule:module];
	if(moduleId == nil)
	{
		@throw [NSException exceptionWithName:NSInvalidArgumentException reason:@"Module has not been registered" userInfo:nil];
	}
	[self removeAllListenersFromModuleWithID:moduleId.intValue forEvent:event];
}

-(void)removeAllListenersFromModuleWithID:(int)moduleId forEvent:(NSString*)event {
	[[self moduleEventsForID:moduleId] removeAllListenersForEvent:event];
}

-(void)removeAllListenersFromModule:(id<RNEventConformer>)module {
	NSNumber* moduleId = [self registeredModuleIDForModule:module];
	if(moduleId == nil)
	{
		@throw [NSException exceptionWithName:NSInvalidArgumentException reason:@"Module has not been registered" userInfo:nil];
	}
	[self removeAllListenersFromModuleWithID:moduleId.intValue];
}

-(void)removeAllListenersFromModuleWithID:(int)moduleId {
	[[self moduleEventsForID:moduleId] removeAllListeners];
}

@end
